src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java

```diff
@@ -0,0 +1,471 @@
+ * that contains the fully-qualified name of your implementation class (or one class per line if
+ * providing multiple implementations).
+ *
+ * <p>If the Dialect implementation is included in this project, the class and the service provider
+ * file will automatically be included in the JDBC Source connector's JAR file. However, it is also
+ * possible to create an implementation outside of this project and to package the dialect (s) and
+ * service provider file in a JAR file, and to then include that JAR file inside the JDBC
+ * Connector's plugin directory within a Connect installation.
+ *
+ * <p>This discovery and registration process uses DEBUG messages to report the {@link
+ * DatabaseDialectProvider} classes that are found and registered. If you have difficulties getting
+ * the connector to find and register your dialect implementation classes, check that your JARs have
+ * the service provider file and your JAR is included in the JDBC connector's plugin directory.
+ */
+public interface DatabaseDialect extends ConnectionProvider {
+
+  /**
+   * Create a new prepared statement using the specified database connection.
+   *
+   * @param db    the database connection; may not be null
+   * @param query the query expression for the prepared statement; may not be null
+   * @return a new prepared statement; never null
+   * @throws SQLException if there is an error with the database connection
+   */
+  PreparedStatement createPreparedStatement(
+      Connection db,
+      String query
+  ) throws SQLException;
+
+  /**
+   * Parse the supplied simple name or fully qualified name for a table into a {@link TableId}.
+   *
+   * @param fqn the fully qualified string representation; may not be null
+   * @return the table identifier; never null
+   */
+  TableId parseTableIdentifier(String fqn);
+
+  /**
+   * Get the identifier rules for this database.
+   *
+   * @return the identifier rules
+   */
+  IdentifierRules identifierRules();
+
+  /**
+   * Get a new {@link ExpressionBuilder} that can be used to build expressions with quoted
+   * identifiers.
+   *
+   * @return the builder; never null
+   * @see #identifierRules()
+   * @see IdentifierRules#expressionBuilder()
+   */
+  ExpressionBuilder expressionBuilder();
+
+  /**
+   * Return current time at the database
+   *
+   * @param conn database connection
+   * @param cal  calendar
+   * @return the current time at the database
+   * @throws SQLException if there is an error with the database connection
+   */
+  Timestamp currentTimeOnDB(
+      Connection conn,
+      Calendar cal
+  ) throws SQLException, ConnectException;
+
+  /**
+   * Get a list of non-system table names in the database.
+   *
+   * @param conn database connection
+   * @return a list of tables; never null
+   * @throws SQLException if there is an error with the database connection
+   */
+  List<TableId> tableNames(
+      Connection conn
+  ) throws SQLException;
```

Comment:
One of the challenges dealing with many different JDBC drivers is that they use different combinations of "catalog", "schema", and "table" in their fully-qualified names. To properly deal with these, you can't simply deal with table names as we were before (there are quite a few issues related to simple names not uniquely identifying tables within a database), so I've introduced `TableId` and `ColumnId` to represent fully-qualified table names and column names, and they can deal with any combination of catalog, schema, and table names.